### PR TITLE
Data integrity project

### PR DESCRIPTION
**Pull request by mthambipillai**

### CODE REVIEW

File: src/java/com/unifina/data/StreamrBinaryMessage.java [hunk 1]
```diff
@@ -2,52 +2,100 @@
...
 			streamIdAsBytes = new byte[streamIdLength];
 			bb.get(streamIdAsBytes);
-			streamId = new String(streamIdAsBytes, StandardCharsets.UTF_8);
+			streamId = new String(streamIdAsBytes, utf8);
 			partition = bb.get() & 0xff; // unsigned byte
 			contentType = bb.get();
 			int contentLength = bb.getInt();
 			content = new byte[contentLength];
 			bb.get(content);
+			if (version == VERSION_SIGNED) {
```
Comment (hpihkala) [inline src/java/com/unifina/data/StreamrBinaryMessage.java]: It's foreseeable that we'll have at least 2 further version increments in the short-to-medium term:

- New message id (version 30)
- Encryption support (version 31)

I'm a little worried about turning this class into a pile of spaghetti with a lot of `if (version == MY_COOL_NEW_VERSION)` branches everywhere. Is there an alternative approach?

How about some more object-oriented style? Version 29 is clearly an extension of 28. There could also be an abstract parent class for shared functionality and bootstrapping the instance creation with a static factory method `from(ByteBuffer)` in this parent class. A few pointers below, I'll leave it to you to play around with the idea if you like it:

```
public abstract class StreamrBinaryMessage {
...
    // There could be a shared version-aware entry point, with the assumption that first byte is version byte
    public static StreamrBinaryMessage from(ByteBuffer bb) {
        byte version = bb.get();
        if (version == 28) {
            return new StreamrBinaryBufferV28(bb);
        } else if (version == 29) {
            return new StreamrBinaryBufferV29(bb);
        } else {
            // throw
        }
    }
}
```
```
public class StreamrBinaryMessageV28 extends StreamrBinaryMessage {
    // Pretty much what the old StreamrBinaryMessage was
}
```
```
public class StreamrBinaryMessageV29 extends StreamrBinaryMessageV28 {
    // New fields only here
	private final byte signatureType;
	private final byte[] addressBytes;
	private final byte[] signatureBytes;

    // Examples of cool stuff the inheritance gives you
    public StreamrBinaryMessageV29(ByteBuffer bb) {
        super(bb);

        // Only read new fields here
        signatureType = bb.get();
	if (signatureType == SIGNATURE_TYPE_ETH) {
		addressBytes = new byte[20];
		bb.get(addressBytes);
		signatureBytes = new byte[65];
		bb.get(signatureBytes);
	} else if (signatureType == SIGNATURE_TYPE_NONE) {
		addressBytes = null;
		signatureBytes = null;
	} else {
		throw new IllegalArgumentException("Unknown signature type: "+signatureType);
	}
    }

    // Another nice perk
    @Override
    public int sizeInBytes() {
        return super.sizeInBytes() + (signatureType == SIGNATURE_TYPE_ETH ? 86 : 1);
    }
}
```

File: src/java/com/unifina/feed/StreamrMessage.java [hunk 2]
```diff
@@ -3,17 +3,33 @@
...
 import java.util.Map;
 
+import com.unifina.data.StreamrBinaryMessage;
 import com.unifina.feed.map.MapMessage;
 
 public class StreamrMessage extends MapMessage {
 
 	private final int partition;
 	private final String streamId;
+	private final byte signatureType;
```
Comment (hpihkala) [inline src/java/com/unifina/feed/StreamrMessage.java]: `StreamrMessage` should not know anything about the byte representation (or `StreamrBinaryMessage`) - perhaps we should make this an enum instead?

File: grails-app/migrations/core/2018-11-27-add-requireSignedData-field.groovy [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+		addColumn(tableName: "stream") {
+			column(name: "require_signed_data", type: "bit") {
+				constraints(nullable: "false")
+			}
+		}
+	}
+	changeSet(author: "mthambipillai", id: "add-requireSignedData-field-3") {
+		dropTable(tableName: "challenge")
+	}
+	changeSet(author: "mthambipillai", id: "add-requireSignedData-field-2") {
```
[the remark this answers is not in the document]
Comment (mthambipillai) [reply grails-app/migrations/core/2018-11-27-add-requireSignedData-field.groovy]: Doing this resulted in the following error in the integration tests:
```
Error Fatal error running tests: Migration failed for change set core/2018-07-12-canvas-started-by.groovy::canvas-started-by-3::mthambipillai:
     Reason: liquibase.exception.DatabaseException: Error executing SQL ALTER TABLE `canvas` ADD CONSTRAINT `FKAE7A7558BA6E1FE8` FOREIGN KEY (`started_by_id`) REFERENCES `sec_user` (`id`): Can't write; duplicate key in table '#sql-1_9':
          Caused By: Error executing SQL ALTER TABLE `canvas` ADD CONSTRAINT `FKAE7A7558BA6E1FE8` FOREIGN KEY (`started_by_id`) REFERENCES `sec_user` (`id`): Can't write; duplicate key in table '#sql-1_9':
          Caused By: Can't write; duplicate key in table '#sql-1_9'
```

File: grails-app/domain/com/unifina/domain/data/Stream.groovy [hunk 4]
```diff
@@ -26,6 +26,8 @@ class Stream implements Comparable {
 	String uiChannelPath
 	Canvas uiChannelCanvas
 
+	Boolean requireSignedData = false
```
[the remark this answers is not in the document]
Comment (hpihkala) [reply grails-app/domain/com/unifina/domain/data/Stream.groovy]: Works for me.

**Pull request merged**